### DISCUSSION TITLE
using Spot SDK in Webots

### DISCUSSION THREAD

**Discussion by MohaKhalili** [cyberbotics/webots#2677] [Ideas]
Hi guys 

I want to suggest by asking some questions :)
Have you heard anything about Boston Dynamics? I think so :D
But do you know about Spot SDK?

If you are a roboticist, chances are you have always wanted to tinker with any of Boston Dynamics’ robots. The Softbank-owned company trying to make those dreams come true by making its Spot SDK publicly available on GitHub. This will allow developers to build custom applications that enable Spot to do useful tasks across a variety of industries.

![screen-shot-2019-09-24-at-9 14 18-am-2048x1280](https://user-images.githubusercontent.com/25219168/105202988-cfb59b80-5b57-11eb-9633-10009ddaffa1.png)
Sample code from the Spot SDK. | Credit: Boston Dynamics

Boston Dynamics shared the following examples of what can be done with the Spot SDK:
- Creating VR-based controls for Spot
- Automating registration of laser-scanning
- Connecting Spot’s data to cloud work order services
- Using edge computing to help Spot semantically understand its environment

Developers who want to actually test their code will need to become part of the Spot Early Adopters Program and lease the quadruped. But there is a catch, Researchers who want to work on robotics fields individually or based on the university for their thesis and have no budget to lease Spot, cannot use Spot SDK. 

We all know that Webots recently added Spot to robots model, how nice it would be if we could use Spot SDK in Webots without any Spot leasing.

From what I´ve chatted with @lukicdarkoo in Discord I think this would be possible to do, actually, I would really like to help to contribute to this.

If you were interested, you can join the discussion, like it and suggest new ideas to convince the community to create the integration.

Best Regards,
Mohammad Khalili

**Comment by omichel** [comment]
I just proposed to add a paragraph about this question in the Webots user guide: https://github.com/cyberbotics/webots/pull/2683
Let me know what you think about it.

**Reply by MohaKhalili** [reply]
I apologize for not answering for a long time 
What does this paragraph describe? 
Can I still say my opinion?

**Reply by omichel** [reply]
See details [here](https://github.com/cyberbotics/webots/pull/2683/files). For sure your opinion is welcome if it is constructive.

**Comment by lukicdarkoo** [comment]
From my point of view, implementing the walking algorithm in the same way as Boston Dynamics did is very hard, as well as getting binaries that we can plug into the simulation.

Our best bet may be to somehow "cheat". We can implement a relatively simple walking algorithm (like this https://github.com/chvmp/champ), and use the Supervisor node to provide additional stability that Spot Mini has.

Anyhow, it is a looot of work! If you think the "cheating" is possible we can put a bounty on this.

**Reply by MohaKhalili** [reply]
I think you are right, but what do you think about Webots partnership with Boston Dynamics? If so, Boston Dynamics will probably give you everything you need to use SDK in Webots as well as getting binaries.
If it is not possible to have a partnership with Boston Dynamics, we can think of other ways like implementing a relatively simple walking algorithm.

**Reply by omichel** [reply]
Setting up such a partnership is a something long, difficult and unlikely to succeed. I would rather recommend you to try to implementation a simple walking algorithm.

**Comment by omichel** [comment]
There are some open-source libraries available to make the Spot walk on flat surfaces and even climb-up stairs: A nice [blog article](https://www.softserveinc.com/en-us/blog/spot-simulation-tools) describes various possibilities. It seems the [MIT Cheetah control algorithm](https://dspace.mit.edu/handle/1721.1/85490) with [champ](https://github.com/chvmp/champ) can be used to make the robot walk on flat surface while [towr](https://github.com/ethz-adrl/towr) will allow the robot to walk on uneven surfaces, climb-up stairs, jump, etc.

**Reply by grassjelly** [reply]
Hi @omichel you might find this repo useful in getting champ able to climb stairs https://github.com/hgaren/aliengo_delivery

**Comment by METEORITENMAX** [comment]
Hi, we are working on a simulation for the spot. The spot can already be controlled via the supplied teleop.

webots_ros2_spot: https://github.com/MASKOR/webots_ros2_spot
